packages/numpy/test_numpy.py

```diff
@@ -191,3 +194,115 @@ def test_runwebworker_numpy(selenium_standalone):
+                    console.log(`${v1}, ${typeof(v1)}, ${v2}, ${typeof(v2)}, ${v1===v2}`);
+                    if(v1.toString() !== v2){
+                        throw new Error(`Discrepancy ${x}[${idx1}, ${idx2}]: ${v1} != ${v2}`);
+                    }
+                }
+            }
+            z.release();
+        }
+        """
+    )
+
+
+@pytest.mark.parametrize(
+    "arg",
+    [
+        "np.arange(6).reshape((2, -1))",
+        pytest.param(
+            "np.arange(12).reshape((3, -1))[::2, ::2]", marks=pytest.mark.xfail
+        ),
+        "np.arange(6).reshape((2, -1)).astype(np.int8, order='C')",
+        "np.arange(6).reshape((2, -1)).astype(np.int8, order='F')",
+        "np.arange(6).reshape((2, -1, 1))",
+        "np.ones((1, 1))[0:0]",  # shape[0] == 0
+        "np.ones(1)",  # ndim == 0
+    ]
+    + [
+        f"np.arange(3).astype(np.{type_})"
+        for type_ in ["int8", "uint8", "int16", "int32", "float32", "float64"]
+    ],
+)
+def test_get_buffer_roundtrip(selenium, arg):
+    selenium.run_js(
+        f"""
+        await pyodide.runPythonAsync(`
+            import numpy as np
+            x = {arg}
+        `);
+        window.x_js_buf = pyodide.pyimport("x").getBuffer();
+        """
+    )
+
+    selenium.run_js(
+        """
+        pyodide.runPython(`
+            from unittest import TestCase
+            from js import x_js_buf
+            assert_equal = TestCase().assertEqual
+
+            assert_equal(x_js_buf.ndim, x.ndim)
+            # The following check fails
+            # assert_equal(x_js_buf.shape.to_py(), list(x.shape))
+            assert_equal(x_js_buf.strides.to_py(), list(x.data.strides))
+            assert_equal(x_js_buf.format, x.data.format)
+            assert_equal(len(x_js_buf.data), np.prod(x.shape))
+            assert_equal(
+                x_js_buf.data.tobytes(order='A').hex(),
+                x.data.tobytes(order='A').hex()
+            )
+            x_js_buf.release()
+        `);
+        """
+    )
+
+
+def test_get_buffer_error_messages(selenium):
+    with pytest.raises(Exception, match="Javascript has no Float16Array"):
+        selenium.run_js(
+            """
+            await pyodide.runPythonAsync(`
+                import numpy as np
+                x = np.ones(2, dtype=np.float16)
+            `);
+            pyodide.pyimport("x").getBuffer();
+            """
+        )
+
+
+@pytest.mark.xfail(reason="should likely fail with a meaninful error message")
+def test_get_buffer_big_endian(selenium):
+    selenium.run_js(
+        """
+        await pyodide.runPythonAsync(`
+            x = np.ones(2, dtype=np.float32).byteswap()
+        `);
+        window.x_js_buf = pyodide.pyimport("x").getBuffer();
+        pyodide.runPython(`
+            from unittest import TestCase
+            from js import x_js_buf
+            assert_equal = TestCase().assertEqual
+            assert_equal(x_js_buf.data[0], 1.0)
```

Comment:
This runs but produces incorrect result. Maybe we should just error on big endian for now?

Reply:
It doesn't seem like `byteswap` changes the format of the array to mark it as having changed endianness. If the test is changed to `assert_equal(x_js_buf.data[0], x[0])` it will pass.

Reply:
Okay I think that these wrong endian buffers are basically impossible to produce and nothing seems to support them. In fact, I think that all of this convoluted code I wrote was unnecessary and the formats can only be single letters. Would be nice if the docs were a little bit easier to understand.

Reply:
Okay, so I removed this test and any mention of endianness from the code.

Reply:
They should be possible to create. Maybe we are doing something wrong. Anyway, as a initial version this is likely fine, let's see if there are user reports about it.

Reply:
Yeah if you can find code that generates a numpy array with `array.data.format == '>I'` or even with `len(array.data.format) > 1` please lmk.